esmf_regrid/esmf_regridder.py

```diff
@@ -191,7 +196,7 @@ def regrid(self, src_array, norm_type="fracarea", mdtol=1):
         normalisations = ma.array(normalisations, mask=np.logical_not(tgt_mask))
 
         flat_src = self.src._array_to_matrix(ma.filled(src_array, 0.0))
-        flat_tgt = self.weight_matrix @ flat_src
+        flat_tgt = self.weight_matrix.astype(out_dtype) @ flat_src
```

Comment:
Although I'm pretty happy with the other code changes,
I'm not so sure whether **_this_** change is wise in addition ? ...

I note that `flat_src` is derived directly from `source_array`, and _will_ always have the same dtype.

I **_assume_** that the '@' type matrix multiply would behave the same as other "ordinary" arithmetic operations, i.e. the result type is always the higher precision of the two.
However, whether that will always 'uprate' either arg to the precision of the other before the calculation, as this is potentially doing. is rather less certain, I would guess -- we _might_ have specific ways to multiply float32 * float64 without pre-converting the lower-precision array. 

So, I would prefer to _trust_ numpy to handle the operation between these different types appropriately, and not to second-guess it.

Meanwhile, there is also a possibility here that weights is a **_higher_** precision than source_array.  In that case, this line would throw that away.
I think that would also be a mistake :  If I supply a higher-precision weights array, then I'd **_expect_** the operation to perform in higher precision in that case, and to return that higher type.

I think that also alters the scope of the change description here : Don't say that it 'preserves dtype', because I don't think that is always the right answer.  The 'correct' result dtype should be the higher of the weights + and the source-array (and not spuriously increased to float64, as was happening).  In any case, an integral input type is not "preserved".

Reply:
OK, I'm convinced that this now works as expected !